Index: zabbix/frontends/php/include/classes/core/CConfigFile.php
===================================================================
--- zabbix.orig/frontends/php/include/classes/core/CConfigFile.php	2016-01-23 13:09:07.190942069 +0100
+++ zabbix/frontends/php/include/classes/core/CConfigFile.php	2016-01-23 13:09:07.186942050 +0100
@@ -24,7 +24,7 @@
 	const CONFIG_NOT_FOUND = 1;
 	const CONFIG_ERROR = 2;
 
-	const CONFIG_FILE_PATH = '/conf/zabbix.conf.php';
+	const CONFIG_FILE_PATH = '/etc/zabbix/zabbix.conf.php';
 
 	private static $supported_db_types = [
 		ZBX_DB_DB2 => true,
Index: zabbix/frontends/php/include/classes/core/ZBase.php
===================================================================
--- zabbix.orig/frontends/php/include/classes/core/ZBase.php	2016-01-23 13:09:07.190942069 +0100
+++ zabbix/frontends/php/include/classes/core/ZBase.php	2016-01-23 13:09:07.186942050 +0100
@@ -299,7 +299,8 @@
 	 * Load zabbix config file.
 	 */
 	protected function loadConfigFile() {
-		$configFile = $this->getRootDir().CConfigFile::CONFIG_FILE_PATH;
+		#$configFile = $this->getRootDir().CConfigFile::CONFIG_FILE_PATH;
+		$configFile = CConfigFile::CONFIG_FILE_PATH;
 		$config = new CConfigFile($configFile);
 		$this->config = $config->load();
 	}
Index: zabbix/frontends/php/include/classes/setup/CSetupWizard.php
===================================================================
--- zabbix.orig/frontends/php/include/classes/setup/CSetupWizard.php	2016-01-23 13:09:07.190942069 +0100
+++ zabbix/frontends/php/include/classes/setup/CSetupWizard.php	2016-01-23 13:09:07.186942050 +0100
@@ -348,8 +348,9 @@
 	function stage5() {
 		$this->setConfig('ZBX_CONFIG_FILE_CORRECT', true);
 
-		$config_file_name = Z::getInstance()->getRootDir().CConfigFile::CONFIG_FILE_PATH;
-		$config = new CConfigFile($config_file_name);
+		#$config_file_name = Z::getInstance()->getRootDir().CConfigFile::CONFIG_FILE_PATH;
+		#$config = new CConfigFile($config_file_name);
+		$config = CConfigFile::CONFIG_FILE_PATH;
 		$config->config = [
 			'DB' => [
 				'TYPE' => $this->getConfig('DB_TYPE'),
@@ -517,7 +518,8 @@
 				// make zabbix.conf.php downloadable
 				header('Content-Type: application/x-httpd-php');
 				header('Content-Disposition: attachment; filename="'.basename(CConfigFile::CONFIG_FILE_PATH).'"');
-				$config = new CConfigFile(Z::getInstance()->getRootDir().CConfigFile::CONFIG_FILE_PATH);
+				#$config = new CConfigFile(Z::getInstance()->getRootDir().CConfigFile::CONFIG_FILE_PATH);
+				$config = new CConfigFile(CConfigFile::CONFIG_FILE_PATH);
 				$config->config = [
 					'DB' => [
 						'TYPE' => $this->getConfig('DB_TYPE'),
