Index: zabbix/frontends/php/include/classes/core/CConfigFile.php
===================================================================
--- zabbix.orig/frontends/php/include/classes/core/CConfigFile.php	2014-09-13 15:11:44.408777846 +0300
+++ zabbix/frontends/php/include/classes/core/CConfigFile.php	2014-09-13 15:11:44.404777827 +0300
@@ -24,7 +24,7 @@
 	const CONFIG_NOT_FOUND = 1;
 	const CONFIG_ERROR = 2;
 
-	const CONFIG_FILE_PATH = '/conf/zabbix.conf.php';
+	const CONFIG_FILE_PATH = '/etc/zabbix/zabbix.conf.php';
 
 	public $configFile = null;
 	public $config = array();
Index: zabbix/frontends/php/include/classes/core/ZBase.php
===================================================================
--- zabbix.orig/frontends/php/include/classes/core/ZBase.php	2014-09-13 15:11:44.408777846 +0300
+++ zabbix/frontends/php/include/classes/core/ZBase.php	2014-09-13 15:11:44.408777846 +0300
@@ -305,7 +305,8 @@
 	 * Load zabbix config file.
 	 */
 	protected function loadConfigFile() {
-		$configFile = $this->getRootDir().CConfigFile::CONFIG_FILE_PATH;
+		#$configFile = $this->getRootDir().CConfigFile::CONFIG_FILE_PATH;
+		$configFile = CConfigFile::CONFIG_FILE_PATH;
 		$config = new CConfigFile($configFile);
 		$this->config = $config->load();
 	}
Index: zabbix/frontends/php/include/classes/setup/CSetupWizard.php
===================================================================
--- zabbix.orig/frontends/php/include/classes/setup/CSetupWizard.php	2014-09-13 15:11:44.408777846 +0300
+++ zabbix/frontends/php/include/classes/setup/CSetupWizard.php	2014-09-13 15:11:44.408777846 +0300
@@ -443,7 +443,8 @@
 	function stage6() {
 		$this->setConfig('ZBX_CONFIG_FILE_CORRECT', true);
 
-		$config = new CConfigFile(Z::getInstance()->getRootDir().CConfigFile::CONFIG_FILE_PATH);
+		#$config = new CConfigFile(Z::getInstance()->getRootDir().CConfigFile::CONFIG_FILE_PATH);
+		$config = CConfigFile::CONFIG_FILE_PATH);
 		$config->config = array(
 			'DB' => array(
 				'TYPE' => $this->getConfig('DB_TYPE'),
@@ -511,7 +512,8 @@
 		$this->HIDE_CANCEL_BUTTON = !$this->DISABLE_NEXT_BUTTON;
 
 
-		$table = array('Configuration file', BR(), '"'.Z::getInstance()->getRootDir().CConfigFile::CONFIG_FILE_PATH.'"',
+		#$table = array('Configuration file', BR(), '"'.Z::getInstance()->getRootDir().CConfigFile::CONFIG_FILE_PATH.'"',
+		$table = array('Configuration file', BR(), '"'.CConfigFile::CONFIG_FILE_PATH.'"',
 			BR(), 'created: ', $this->getConfig('ZBX_CONFIG_FILE_CORRECT', false)
 			? new CSpan(_('OK'), 'ok')
 			: new CSpan(_('Fail'), 'fail')
@@ -524,7 +526,7 @@
 				? array($error_text, BR(), 'Please install it manually, or fix permissions on the conf directory.', BR(), BR(),
 					'Press the "Download configuration file" button, download the configuration file ',
 					'and save it as ', BR(),
-					'"'.Z::getInstance()->getRootDir().CConfigFile::CONFIG_FILE_PATH.'"', BR(), BR(),
+					'"'.CConfigFile::CONFIG_FILE_PATH.'"', BR(), BR(),
 					new CSubmit('save_config', 'Download configuration file'),
 					BR(), BR()
 				)
@@ -635,7 +637,8 @@
 				// make zabbix.conf.php downloadable
 				header('Content-Type: application/x-httpd-php');
 				header('Content-Disposition: attachment; filename="'.basename(CConfigFile::CONFIG_FILE_PATH).'"');
-				$config = new CConfigFile(Z::getInstance()->getRootDir().CConfigFile::CONFIG_FILE_PATH);
+				#$config = new CConfigFile(Z::getInstance()->getRootDir().CConfigFile::CONFIG_FILE_PATH);
+				$config = new CConfigFile(CConfigFile::CONFIG_FILE_PATH);
 				$config->config = array(
 					'DB' => array(
 						'TYPE' => $this->getConfig('DB_TYPE'),
