Index: zabbix/src/zabbix_java/Makefile.in
===================================================================
--- zabbix.orig/src/zabbix_java/Makefile.in	2014-09-13 15:11:43.632774299 +0300
+++ zabbix/src/zabbix_java/Makefile.in	2014-09-13 15:11:43.628774281 +0300
@@ -490,6 +490,7 @@
 class:
 	$(MKDIR_P) class/src
 	$(MKDIR_P) class/tests
+	$(MKDIR_P) bin
 
 install: all
 	$(MKDIR_P) "$(ZJG_DEST)/bin"
Index: zabbix/src/zabbix_java/Makefile.am
===================================================================
--- zabbix.orig/src/zabbix_java/Makefile.am	2014-09-13 15:11:43.632774299 +0300
+++ zabbix/src/zabbix_java/Makefile.am	2014-09-13 15:11:43.628774281 +0300
@@ -28,6 +28,7 @@
 class:
 	$(MKDIR_P) class/src
 	$(MKDIR_P) class/tests
+	$(MKDIR_P) bin
 
 install: all
 	$(MKDIR_P) "$(ZJG_DEST)/bin"
