--- a/src/zabbix_java/Makefile.in
+++ b/src/zabbix_java/Makefile.in
@@ -503,6 +503,7 @@ test: class
 class:
 	$(MKDIR_P) class/src
 	$(MKDIR_P) class/tests
+	$(MKDIR_P) bin
 
 install: all
 	$(MKDIR_P) "$(ZJG_DEST)/bin"
--- a/src/zabbix_java/Makefile.am
+++ b/src/zabbix_java/Makefile.am
@@ -28,6 +28,7 @@ test: class
 class:
 	$(MKDIR_P) class/src
 	$(MKDIR_P) class/tests
+	$(MKDIR_P) bin
 
 install: all
 	$(MKDIR_P) "$(ZJG_DEST)/bin"
