Index: zabbix/src/zabbix_java/Makefile.in
===================================================================
--- zabbix.orig/src/zabbix_java/Makefile.in	2015-02-23 21:14:34.381788609 +0100
+++ zabbix/src/zabbix_java/Makefile.in	2015-02-23 21:14:34.377788609 +0100
@@ -490,6 +490,7 @@
 class:
 	$(MKDIR_P) class/src
 	$(MKDIR_P) class/tests
+	$(MKDIR_P) bin
 
 install: all
 	$(MKDIR_P) "$(ZJG_DEST)/bin"
Index: zabbix/src/zabbix_java/Makefile.am
===================================================================
--- zabbix.orig/src/zabbix_java/Makefile.am	2015-02-23 21:14:34.381788609 +0100
+++ zabbix/src/zabbix_java/Makefile.am	2015-02-23 21:14:34.377788609 +0100
@@ -28,6 +28,7 @@
 class:
 	$(MKDIR_P) class/src
 	$(MKDIR_P) class/tests
+	$(MKDIR_P) bin
 
 install: all
 	$(MKDIR_P) "$(ZJG_DEST)/bin"
