Last-Update: 2012-07-17
Author: Dmitry Smirnov <onlyjob@member.fsf.org>
Forwarded: not-needed
Description: correct location of config files in man pages

Index: zabbix/man/zabbix_agentd.man
===================================================================
--- zabbix.orig/man/zabbix_agentd.man	2016-01-23 13:09:06.870940518 +0100
+++ zabbix/man/zabbix_agentd.man	2016-01-23 13:09:06.866940499 +0100
@@ -116,7 +116,7 @@
 Output version information and exit.
 .SH FILES
 .TP
-.I /usr/local/etc/zabbix_agentd.conf
+.I /etc/zabbix/zabbix_agentd.conf
 Default location of Zabbix agent configuration file (if not modified during compile time).
 .SH "SEE ALSO"
 .BR zabbix_get (8),
Index: zabbix/man/zabbix_proxy.man
===================================================================
--- zabbix.orig/man/zabbix_proxy.man	2016-01-23 13:09:06.870940518 +0100
+++ zabbix/man/zabbix_proxy.man	2016-01-23 13:09:06.866940499 +0100
@@ -79,7 +79,7 @@
 Output version information and exit.
 .SH FILES
 .TP
-.I /usr/local/etc/zabbix_proxy.conf
+.I /etc/zabbix/zabbix_proxy.conf
 Default location of Zabbix proxy configuration file (if not modified during compile time).
 .SH "SEE ALSO"
 .BR zabbix_agentd (8),
Index: zabbix/man/zabbix_server.man
===================================================================
--- zabbix.orig/man/zabbix_server.man	2016-01-23 13:09:06.870940518 +0100
+++ zabbix/man/zabbix_server.man	2016-01-23 13:09:06.866940499 +0100
@@ -78,7 +78,7 @@
 Output version information and exit.
 .SH FILES
 .TP
-.I /usr/local/etc/zabbix_server.conf
+.I /etc/zabbix/zabbix_server.conf
 Default location of Zabbix server configuration file (if not modified during compile time).
 .SH "SEE ALSO"
 .BR zabbix_agentd (8),
Index: zabbix/man/zabbix_sender.man
===================================================================
--- zabbix.orig/man/zabbix_sender.man	2016-01-23 13:09:06.870940518 +0100
+++ zabbix/man/zabbix_sender.man	2016-01-23 13:09:06.866940499 +0100
@@ -436,7 +436,7 @@
 .RE
 
 .br
-\fBecho "\- hw.serial.number 1287872261 SQ4321ASDF" | zabbix_sender \-c /usr/local/etc/zabbix_agentd.conf \-T \-i \-\fR
+\fBecho "\- hw.serial.number 1287872261 SQ4321ASDF" | zabbix_sender \-c /etc/zabbix/zabbix_agentd.conf \-T \-i \-\fR
 .br
 .RS
 Send a timestamped value from the commandline to Zabbix server, specified in the agent configuration file.
